File: android/src/main/java/com/easemob/im_flutter_sdk/ClientWrapper.java
package com.easemob.im_flutter_sdk;

import java.util.Map;
import java.util.HashMap;
import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

import com.hyphenate.EMConnectionListener;
import com.hyphenate.chat.EMClient;
import com.hyphenate.chat.EMOptions;

import com.hyphenate.exceptions.HyphenateException;


import org.json.JSONException;
import org.json.JSONObject;


public class ClientWrapper extends Wrapper implements MethodCallHandler {

    private ChatManagerWrapper chatManagerWrapper;
    private ChatRoomManagerWrapper chatRoomManagerWrapper;
    private ConversationWrapper conversationWrapper;
    private MessageWrapper messageWrapper;
    public ProgressManager progressManager;
    private EMConnectionListener connectionListener;

    private EMOptions options;

    ClientWrapper(FlutterPlugin.FlutterPluginBinding flutterPluginBinding, String channelName) {
        super(flutterPluginBinding, channelName);
    }

    public void sendDataToFlutter(final Map data) {
        if (data == null) {
            return;
        }
        post(()-> channel.invokeMethod(MethodKey.onSendDataToFlutter, data));
    }

    @Override
    public void onMethodCall(MethodCall call, @NonNull Result result) {

        JSONObject param = (JSONObject)call.arguments;
        try {
            if (MethodKey.init.equals(call.method)) {
                init(param, call.method, result);
            }
            else if (MethodKey.createAccount.equals(call.method))
            {
                createAccount(param, call.method, result);
            }
            else if (MethodKey.login.equals(call.method))
            {
                login(param, call.method, result);
            }
            else if (MethodKey.logout.equals(call.method))
            {
                logout(param, call.method, result);
            }
            else if (MethodKey.uploadLog.equals(call.method))
            {
                uploadLog(param, call.method, result);
            }
            else if (MethodKey.compressLogs.equals(call.method))
            {
                compressLogs(param, call.method, result);
            }
            else if (MethodKey.isLoggedInBefore.equals(call.method))
            {
                isLoggedInBefore(param, call.method, result);
            }
            else if (MethodKey.getCurrentUser.equals(call.method))
            {
                getCurrentUser(param, call.method, result);
            }
            else if (MethodKey.getToken.equals(call.method))
            {
                getToken(param, call.method, result);
            }
            else if (MethodKey.isConnected.equals(call.method)) {
                isConnected(param, call.method, result);
            }
            else if (MethodKey.renewToken.equals(call.method)){
                renewToken(param, call.method, result);
            } else if (MethodKey.startCallback.equals(call.method)) {
                startCallback(param, call.method, result);
            }
            // 481
            else if (MethodKey.updateUsingHttpsOnlySetting.equals(call.method)) {
                updateUsingHttpsOnlySetting(param, call.method, result);
            }
            else if (MethodKey.updateDeleteMessageWhenLeaveRoomSetting.equals(call.method)) {
                updateDeleteMessageWhenLeaveRoomSetting(param, call.method, result);
            }
            else if (MethodKey.updateRoomOwnerCanLeaveSetting.equals(call.method)) {
                updateRoomOwnerCanLeaveSetting(param, call.method, result);
            }
            else if (MethodKey.updateAutoDownloadAttachmentThumbnailSetting.equals(call.method)) {
                updateAutoDownloadAttachmentThumbnailSetting(param, call.method, result);
            }
            else if (MethodKey.updateRequireAckSetting.equals(call.method)) {
                updateRequireAckSetting(param, call.method, result);
            }
            else if (MethodKey.updateDeliveryAckSetting.equals(call.method)) {
                updateDeliveryAckSetting(param, call.method, result);
            }
            else if (MethodKey.updateSortMessageByServerTimeSetting.equals(call.method)) {
                updateSortMessageByServerTimeSetting(param, call.method, result);
            }
            else  {
                super.onMethodCall(call, result);
            }

        }catch (JSONException ignored) {

        }
    }
    private void createAccount(JSONObject param, String channelName, Result result) throws JSONException {
        String username = param.getString("username");
        String password = param.getString("password");
        asyncRunnable(()->{
            try {
                EMClient.getInstance().createAccount(username, password);
                onSuccess(result, channelName, username);
            } catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void login(JSONObject param, String channelName, Result result) throws JSONException {
        boolean isPwd = param.getBoolean("isPassword");
        String username = param.getString("username");
        String pwdOrToken = param.getString("pwdOrToken");
        EMWrapperCallBack callBack = new EMWrapperCallBack(result, channelName, null) {
            @Override
            public void onSuccess() {
                post(() -> {
                    object = EMClient.getInstance().getCurrentUser();
                    super.onSuccess();
                });
            }
        };

        if (isPwd){
            EMClient.getInstance().login(username, pwdOrToken, callBack);
        } else {
            EMClient.getInstance().loginWithToken(username, pwdOrToken, callBack);
        }
    }


    private void logout(JSONObject param, String channelName, Result result) throws JSONException {
        boolean unbindToken = param.getBoolean("unbindToken");
        EMClient.getInstance().logout(unbindToken, new EMWrapperCallBack(result, channelName, null){
            @Override
            public void onSuccess() {
                ListenerHandle.getInstance().clearHandle();
                object = true;
                super.onSuccess();
            }
        });
    }

    private void getCurrentUser(JSONObject param, String channelName, Result result) throws JSONException {
        asyncRunnable(()-> onSuccess(result, channelName, EMClient.getInstance().getCurrentUser()));
    }

    private void getToken(JSONObject param, String channelName, Result result) throws JSONException
    {
        asyncRunnable(()-> onSuccess(result, channelName, EMClient.getInstance().getAccessToken()));
    }

    private void isLoggedInBefore(JSONObject param, String channelName, Result result) throws JSONException {
        asyncRunnable(()->{
            EMOptions emOptions = EMClient.getInstance().getOptions();
            onSuccess(result, channelName, EMClient.getInstance().isLoggedInBefore() && emOptions.getAutoLogin());
        });
    }

    private void isConnected(JSONObject param, String channelName, Result result) throws JSONException{
        asyncRunnable(()-> onSuccess(result, channelName, EMClient.getInstance().isConnected()));
    }

    private void uploadLog(JSONObject param, String channelName, Result result) throws JSONException {
        EMClient.getInstance().uploadLog(new EMWrapperCallBack(result, channelName, true));
    }

    private void compressLogs(JSONObject param, String channelName, Result result) throws JSONException {
        asyncRunnable(()->{
            try {
                String path = EMClient.getInstance().compressLogs();
                onSuccess(result, channelName, path);
            } catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void init(JSONObject param, String channelName, Result result) throws JSONException {
        if (options != null) {
            onSuccess(result, channelName, null);
            return;
        }
        options = OptionsHelper.fromJson(param, this.context);
        EMClient.getInstance().init(this.context, options);
        EMClient.getInstance().setDebugMode(param.getBoolean("debugModel"));

        bindingManagers();
        registerEaseListener();

        onSuccess(result, channelName, null);

    }

    private void renewToken(JSONObject param, String channelName, Result result) throws JSONException {
        String agoraToken = param.getString("agora_token");
        EMClient.getInstance().renewToken(agoraToken, new EMWrapperCallBack(result, channelName,null));
    }

    private void startCallback(JSONObject param, String channelName, Result result) {
        ListenerHandle.getInstance().startCallback();
        onSuccess(result, channelName, null);
    }

    private void bindingManagers() {
        chatManagerWrapper = new ChatManagerWrapper(binging, "chat_manager");
        chatRoomManagerWrapper = new ChatRoomManagerWrapper(binging, "chat_room_manager");
        conversationWrapper = new ConversationWrapper(binging, "chat_conversation");
        messageWrapper = new MessageWrapper(binging, "chat_message");
        progressManager = new ProgressManager(binging, "file_progress_manager");
    }

    private void clearAllListener() {
        if (chatManagerWrapper != null) chatManagerWrapper.unRegisterEaseListener();
        if (chatRoomManagerWrapper != null) chatRoomManagerWrapper.unRegisterEaseListener();
        if (conversationWrapper != null) conversationWrapper.unRegisterEaseListener();
        if (messageWrapper != null) messageWrapper.unRegisterEaseListener();
        if (progressManager != null) progressManager.unRegisterEaseListener();
    }

    @Override
    public void unRegisterEaseListener() {
        EMClient.getInstance().removeConnectionListener(connectionListener);
        clearAllListener();
    }


    private void registerEaseListener() {


        if (connectionListener != null) {
            EMClient.getInstance().removeConnectionListener(connectionListener);
        }

        connectionListener = new EMConnectionListener() {
            @Override
            public void onConnected() {
                Map<String, Object> data = new HashMap<>();
                data.put("connected", Boolean.TRUE);
                post(()-> channel.invokeMethod(MethodKey.onConnected, data));
            }

            @Override
            public void onDisconnected(int errorCode) {
                if (errorCode == 207) {
                    ListenerHandle.getInstance().clearHandle();
                    post(() -> channel.invokeMethod(MethodKey.onUserDidRemoveFromServer, null));
                }else if (errorCode == 305) {
                    ListenerHandle.getInstance().clearHandle();
                    post(() -> channel.invokeMethod(MethodKey.onUserDidForbidByServer, null));
                }else if (errorCode == 216) {
                    ListenerHandle.getInstance().clearHandle();
                    post(() -> channel.invokeMethod(MethodKey.onUserDidChangePassword, null));
                }else if (errorCode == 214) {
                    ListenerHandle.getInstance().clearHandle();
                    post(() -> channel.invokeMethod(MethodKey.onUserDidLoginTooManyDevice, null));
                }
                else if (errorCode == 217) {
                    ListenerHandle.getInstance().clearHandle();
                    post(() -> channel.invokeMethod(MethodKey.onUserKickedByOtherDevice, null));
                }
                else if (errorCode == 202) {
                    ListenerHandle.getInstance().clearHandle();
                    post(() -> channel.invokeMethod(MethodKey.onUserAuthenticationFailed, null));
                }
                else if (errorCode == 8) {
                    post(() -> channel.invokeMethod(MethodKey.onAppActiveNumberReachLimit, null));
                }
                else {
                    post(() -> channel.invokeMethod(MethodKey.onDisconnected, null));
                }
            }

            @Override
            public void onTokenExpired() {
                post(()-> channel.invokeMethod(MethodKey.onTokenDidExpire, null));
            }

            @Override
            public void onTokenWillExpire() {
                post(()-> channel.invokeMethod(MethodKey.onTokenWillExpire, null));
            }

        };

        EMClient.getInstance().addConnectionListener(connectionListener);

    }

    // 481
    private void updateUsingHttpsOnlySetting(JSONObject param, String channelName, Result result) throws JSONException {
        boolean usingHttpsOnly = param.getBoolean("usingHttpsOnly");
        EMClient.getInstance().getOptions().setUsingHttpsOnly(usingHttpsOnly);
        asyncRunnable(()->onSuccess(result, channelName, null));
    }

    private void updateDeleteMessageWhenLeaveRoomSetting(JSONObject param, String channelName, Result result) throws JSONException {
        boolean deleteMessageWhenLeaveRoom = param.getBoolean("deleteMessageWhenLeaveRoom");
        EMClient.getInstance().getOptions().setDeleteMessagesAsExitChatRoom(deleteMessageWhenLeaveRoom);
        asyncRunnable(()-> onSuccess(result, channelName, null));
    }
    private void updateRoomOwnerCanLeaveSetting(JSONObject param, String channelName, Result result) throws JSONException {
        boolean roomOwnerCanLeave = param.getBoolean("roomOwnerCanLeave");
        EMClient.getInstance().getOptions().allowChatroomOwnerLeave(roomOwnerCanLeave);
        asyncRunnable(()-> onSuccess(result, channelName, null));
    }

    private void updateAutoDownloadAttachmentThumbnailSetting(JSONObject param, String channelName, Result result) throws JSONException {
        boolean autoDownloadThumbnail = param.getBoolean("autoDownloadThumbnail");
        EMClient.getInstance().getOptions().setAutoDownloadThumbnail(autoDownloadThumbnail);
        asyncRunnable(()-> onSuccess(result, channelName, null));
    }
    private void updateRequireAckSetting(JSONObject param, String channelName, Result result) throws JSONException {
        boolean requireAck = param.getBoolean("requireAck");
        EMClient.getInstance().getOptions().setRequireAck(requireAck);
        asyncRunnable(()-> onSuccess(result, channelName, null));
    }
    private void updateDeliveryAckSetting(JSONObject param, String channelName, Result result) throws JSONException {
        boolean requireDeliveryAck = param.getBoolean("requireDeliveryAck");
        EMClient.getInstance().getOptions().setRequireDeliveryAck(requireDeliveryAck);
        asyncRunnable(()-> onSuccess(result, channelName, null));
    }
    private void updateSortMessageByServerTimeSetting(JSONObject param, String channelName, Result result) throws JSONException {
        boolean sortMessageByServerTime = param.getBoolean("sortMessageByServerTime");
        EMClient.getInstance().getOptions().setSortMessageByServerTime(sortMessageByServerTime);
        asyncRunnable(()-> onSuccess(result, channelName, null));
    }

}

File: ios/Classes/ClientWrapper.m
//
//  EMClientWrapper.m
//  
//
//  Created by 杜洁鹏 on 2019/10/8.
//

#import "ClientWrapper.h"
#import "MethodKeys.h"
#import "ChatManagerWrapper.h"
#import "ConversationWrapper.h"
#import "ChatroomManagerWrapper.h"
#import "OptionsHelper.h"
#import "MessageWrapper.h"
#import "ProgressManager.h"
#import "ListenerHandle.h"
#import "EnumTools.h"

@interface ClientWrapper () <EMClientDelegate, FlutterPlugin>
{
    ChatManagerWrapper *_chatManager;
    ConversationWrapper *_conversationManager;
    ChatroomManagerWrapper *_roomManager;
    MessageWrapper *_msgWrapper;
    ProgressManager *_progressManager;
}
@end

@implementation ClientWrapper {
    EMOptions *_options;
}



- (void)sendDataToFlutter:(NSDictionary *)aData {
    if (aData == nil) {
        return;
    }
    [self.channel invokeMethod:ChatSendDataToFlutter
                     arguments:aData];
}


- (instancetype)initWithChannelName:(NSString *)aChannelName
                          registrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    if(self = [super initWithChannelName:aChannelName
                               registrar:registrar]) {
        
    }
    return self;
}

#pragma mark - FlutterPlugin

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([ChatInit isEqualToString:call.method])
    {
        [self initSDKWithDict:call.arguments
                  channelName:call.method
                       result:result];
    }
    else if ([ChatCreateAccount isEqualToString:call.method])
    {
        [self createAccount:call.arguments
                channelName:call.method
                     result:result];
    }
    else if ([ChatLogin isEqualToString:call.method])
    {
        [self login:call.arguments
        channelName:call.method
             result:result];
    }
    else if ([ChatLogout isEqualToString:call.method])
    {
        [self logout:call.arguments
         channelName:call.method
              result:result];
    }
    else if ([ChatUploadLog isEqualToString:call.method])
    {
        [self uploadLog:call.arguments
            channelName:call.method
                 result:result];
    }
    else if ([ChatCompressLogs isEqualToString:call.method])
    {
        [self compressLogs:call.arguments
               channelName:call.method
                    result:result];
    }
    else if([ChatIsLoggedInBefore isEqualToString:call.method])
    {
        [self isLoggedInBefore:call.arguments
                   channelName:call.method
                        result:result];
    }
    else if([ChatGetCurrentUser isEqualToString:call.method])
    {
        [self getCurrentUser:call.arguments
                 channelName:call.method
                      result:result];
    }
    else if([ChatGetToken isEqualToString:call.method])
    {
        [self getToken:call.arguments
           channelName:call.method
                result:result];
    }
    else if([ChatIsConnected isEqualToString:call.method])
    {
        [self isConnected:call.arguments
              channelName:call.method
                   result:result];
    }
    else if ([ChatRenewToken isEqualToString:call.method]){
        [self renewToken:call.arguments
             channelName:call.method
                  result:result];
    }else if ([ChatStartCallback isEqualToString:call.method]){
        [self startCallBack:call.arguments
                channelName:call.method
                     result:result];
    }
    // 481
    else if ([ChatUpdateUsingHttpsOnlySetting isEqualToString:call.method]){
        [self updateUsingHttpsOnlySetting:call.arguments
                              channelName:call.method
                                   result:result];
    }
    else if ([ChatUpdateDeleteMessageWhenLeaveRoomSetting isEqualToString:call.method]){
        [self updateDeleteMessageWhenLeaveRoomSetting:call.arguments
                                          channelName:call.method
                                               result:result];
    }
    else if ([ChatUpdateRoomOwnerCanLeaveSetting isEqualToString:call.method]){
        [self updateRoomOwnerCanLeaveSetting:call.arguments
                                 channelName:call.method
                                      result:result];
    }
    else if ([ChatUpdateAutoDownloadAttachmentThumbnailSetting isEqualToString:call.method]){
        [self updateAutoDownloadAttachmentThumbnailSetting:call.arguments
                                               channelName:call.method
                                                    result:result];
    }
    else if ([ChatUpdateRequireAckSetting isEqualToString:call.method]){
        [self updateRequireAckSetting:call.arguments
                          channelName:call.method
                               result:result];
    }
    else if ([ChatUpdateDeliveryAckSetting isEqualToString:call.method]){
        [self updateDeliveryAckSetting:call.arguments
                           channelName:call.method
                                result:result];
    }
    else if ([ChatUpdateSortMessageByServerTimeSetting isEqualToString:call.method]){
        [self updateSortMessageByServerTimeSetting:call.arguments
                                       channelName:call.method
                                            result:result];
    }
    else {
        [super handleMethodCall:call result:result];
    }
}


#pragma mark - Actions
- (void)initSDKWithDict:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    
    __weak typeof(self) weakSelf = self;
    
    if(_options) {
        [weakSelf wrapperCallBack:result
                      channelName:ChatInit
                            error:nil
                           object:nil];
        return;
    }
    
    _options = [EMOptions fromJson:param];

    [EMClient.sharedClient initializeSDKWithOptions:_options];

    [self registerManagers];
    [weakSelf wrapperCallBack:result
                  channelName:ChatInit
                        error:nil
                       object:nil];
}


- (void)registerManagers {
    [self clearAllListener];
    [EMClient.sharedClient addDelegate:self delegateQueue:nil];
    _chatManager = [[ChatManagerWrapper alloc] initWithChannelName:EMChannelName(@"chat_manager")registrar:self.flutterPluginRegister];
    _conversationManager = [[ConversationWrapper alloc] initWithChannelName:EMChannelName(@"chat_conversation") registrar:self.flutterPluginRegister];
    _roomManager =[[ChatroomManagerWrapper alloc] initWithChannelName:EMChannelName(@"chat_room_manager") registrar:self.flutterPluginRegister];
    _msgWrapper = [[MessageWrapper alloc] initWithChannelName:EMChannelName(@"chat_message") registrar:self.flutterPluginRegister];
    _progressManager = [[ProgressManager alloc] initWithChannelName:EMChannelName(@"file_progress_manager")  registrar:self.flutterPluginRegister];
}

- (void)clearAllListener {
    [_chatManager unRegisterEaseListener];
    [_conversationManager unRegisterEaseListener];
    [_roomManager unRegisterEaseListener];
    [_msgWrapper unRegisterEaseListener];
    [super unRegisterEaseListener];
    [EMClient.sharedClient removeDelegate:self];
}

// 由父类调用，不需要调用 clearAllListener方法，每个manager中都由父类调用。
- (void)unRegisterEaseListener {
    [EMClient.sharedClient removeDelegate:self];
}

- (ProgressManager *)progressManager {
    return _progressManager;
}

- (void)createAccount:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    NSString *username = param[@"username"];
    NSString *password = param[@"password"];
    [EMClient.sharedClient registerWithUsername:username
                                       password:password
                                     completion:^(NSString *aUsername, EMError *aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aUsername];
    }];
}

- (void)login:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    NSString *username = param[@"username"];
    NSString *pwdOrToken = param[@"pwdOrToken"];
    BOOL isPwd = [param[@"isPassword"] boolValue];
    
    if (isPwd) {

        [EMClient.sharedClient loginWithUsername:username
                                        password:pwdOrToken
                                      completion:^(NSString *aUsername, EMError *aError){

            [weakSelf wrapperCallBack:result
                          channelName:aChannelName
                                error:aError
                               object:EMClient.sharedClient.currentUsername];
        }];
    }else {
        [EMClient.sharedClient loginWithUsername:username
                                           token:pwdOrToken
                                      completion:^(NSString *aUsername, EMError *aError)
         {
            [weakSelf wrapperCallBack:result
                          channelName:aChannelName
                                error:aError
                               object:EMClient.sharedClient.currentUsername];
        }];
    }
}

- (void)logout:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    BOOL unbindToken = [param[@"unbindToken"] boolValue];
    [EMClient.sharedClient logout:unbindToken completion:^(EMError *aError) {
        if(aError == nil) {
            [ListenerHandle.sharedInstance clearHandle];
        }
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:@(!aError)];
    }];
}

- (void)getCurrentUser:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    NSString* username = EMClient.sharedClient.currentUsername;
    [weakSelf wrapperCallBack:result
                  channelName:aChannelName
                        error:nil
                       object:username];
    
}

- (void)uploadLog:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    [EMClient.sharedClient uploadDebugLogToServerWithCompletion:^(EMError *aError) {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:nil];
    }];
}

- (void)compressLogs:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    [EMClient.sharedClient getLogFilesPathWithCompletion:^(NSString *aPath, EMError *aError) {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aPath];
    }];
}

- (void)isLoggedInBefore:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    [weakSelf wrapperCallBack:result
                  channelName:aChannelName
                        error:nil
                       object:@(EMClient.sharedClient.isLoggedIn)];
    
}

- (void)getToken:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result{
    [self wrapperCallBack:result
                  channelName:aChannelName
                        error:nil
                       object:EMClient.sharedClient.accessUserToken];
}


- (void)isConnected:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result{
    [self wrapperCallBack:result
                  channelName:aChannelName
                        error:nil
                       object:@(EMClient.sharedClient.isConnected)];
}

- (void)renewToken:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result{
    __weak typeof(self)weakSelf = self;
    NSString *newAgoraToken = param[@"agora_token"];
    [EMClient.sharedClient renewToken:newAgoraToken completion:^(EMError * _Nullable aError) {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:nil
                           object:nil];
    }];
}

- (void)startCallBack:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result{
    __weak typeof(self)weakSelf = self;
    [ListenerHandle.sharedInstance startCallback];
    [weakSelf wrapperCallBack:result
                  channelName:aChannelName
                        error:nil
                       object:nil];
}

#pragma - mark EMClientDelegate

- (void)connectionStateDidChange:(EMConnectionState)aConnectionState {
    BOOL isConnected = aConnectionState == EMConnectionConnected;
    if (isConnected) {
        [self.channel invokeMethod:ChatOnConnected
                         arguments:nil];
    }else {
        [self.channel invokeMethod:ChatOnDisconnected
                         arguments:nil];
    }
}

- (void)autoLoginDidCompleteWithError:(EMError *)aError {
    if (aError.code == EMErrorServerServingForbidden) {
         [self userDidForbidByServer];
    }else if (aError.code == EMErrorAppActiveNumbersReachLimitation) {
        [self activeNumbersReachLimitation];
    }
}

- (void)activeNumbersReachLimitation {
    [self.channel invokeMethod:ChatOnAppActiveNumberReachLimit arguments:nil];
}

// 声网token即将过期
- (void)tokenWillExpire:(EMErrorCode)aErrorCode {
    [self.channel invokeMethod:ChatOnTokenWillExpire arguments:nil];
}

// 声网token过期
- (void)tokenDidExpire:(EMErrorCode)aErrorCode {
    [ListenerHandle.sharedInstance clearHandle];
    [self.channel invokeMethod:ChatOnTokenDidExpire
                     arguments:nil];
}

//- (void)userAccountDidLoginFromOtherDevice:(NSString *)aDeviceName {
//    [EMListenerHandle.sharedInstance clearHandle];
//    [self.channel invokeMethod:ChatOnUserDidLoginFromOtherDevice
//                     arguments:@{@"deviceName": aDeviceName}];
//}

- (void)userAccountDidRemoveFromServer {
    [ListenerHandle.sharedInstance clearHandle];
    [self.channel invokeMethod:ChatOnUserDidRemoveFromServer
                     arguments:nil];
}

- (void)userDidForbidByServer {
    [ListenerHandle.sharedInstance clearHandle];
    [self.channel invokeMethod:ChatOnUserDidForbidByServer
                     arguments:nil];
}


- (void)userAccountDidForcedToLogout:(EMError *)aError {
    [ListenerHandle.sharedInstance clearHandle];
    if (aError.code == EMErrorUserKickedByChangePassword) {
        [self.channel invokeMethod:ChatOnUserDidChangePassword
                         arguments:nil];
    } else if (aError.code == EMErrorUserLoginTooManyDevices) {
        [self.channel invokeMethod:ChatOnUserDidLoginTooManyDevice
                         arguments:nil];
    } else if (aError.code == EMErrorUserKickedByOtherDevice) {
        [self.channel invokeMethod:ChatOnUserKickedByOtherDevice
                         arguments:nil];
    } else if (aError.code == EMErrorUserAuthenticationFailed) {
        [self.channel invokeMethod:ChatOnUserAuthenticationFailed
                         arguments:nil];
    }
}


# pragma mark - 481
- (void)userAccountDidLoginFromOtherDevice:(NSString *)aDeviceName {
    [ListenerHandle.sharedInstance clearHandle];
    [self.channel invokeMethod:ChatOnUserDidLoginFromOtherDevice
                     arguments:aDeviceName];
}


- (void)updateUsingHttpsOnlySetting:(NSDictionary *)param
                        channelName:(NSString *)aChannelName
                             result:(FlutterResult)result 
{
    __weak typeof(self)weakSelf = self;
    BOOL usingHttpsOnly = [param[@"usingHttpsOnly"] boolValue];
    EMClient.sharedClient.options.usingHttpsOnly = usingHttpsOnly;
    [weakSelf wrapperCallBack:result
                  channelName:aChannelName
                        error:nil
                       object:nil];
}

- (void)updateDeleteMessageWhenLeaveRoomSetting:(NSDictionary *)param
                                    channelName:(NSString *)aChannelName
                                         result:(FlutterResult)result
{
    __weak typeof(self)weakSelf = self;
    BOOL deleteMessageWhenLeaveRoom = [param[@"deleteMessageWhenLeaveRoom"] boolValue];
    EMClient.sharedClient.options.deleteMessagesOnLeaveChatroom = deleteMessageWhenLeaveRoom;
    [weakSelf wrapperCallBack:result
                  channelName:aChannelName
                        error:nil
                       object:nil];
}

- (void)updateRoomOwnerCanLeaveSetting:(NSDictionary *)param
                           channelName:(NSString *)aChannelName
                                result:(FlutterResult)result
{
    __weak typeof(self)weakSelf = self;
    BOOL roomOwnerCanLeave = [param[@"roomOwnerCanLeave"] boolValue];
    EMClient.sharedClient.options.canChatroomOwnerLeave = roomOwnerCanLeave;
    [weakSelf wrapperCallBack:result
                  channelName:aChannelName
                        error:nil
                       object:nil];
}

- (void)updateAutoDownloadAttachmentThumbnailSetting:(NSDictionary *)param
                                         channelName:(NSString *)aChannelName
                                              result:(FlutterResult)result
{
    __weak typeof(self)weakSelf = self;
    BOOL autoDownloadThumbnail = [param[@"autoDownloadThumbnail"] boolValue];
    EMClient.sharedClient.options.autoDownloadThumbnail = autoDownloadThumbnail;
    [weakSelf wrapperCallBack:result
                  channelName:aChannelName
                        error:nil
                       object:nil];
}

- (void)updateRequireAckSetting:(NSDictionary *)param
                    channelName:(NSString *)aChannelName
                         result:(FlutterResult)result
{
    __weak typeof(self)weakSelf = self;
    BOOL requireAck = [param[@"requireAck"] boolValue];
    EMClient.sharedClient.options.enableRequireReadAck = requireAck;
    [weakSelf wrapperCallBack:result
                  channelName:aChannelName
                        error:nil
                       object:nil];
}

- (void)updateDeliveryAckSetting:(NSDictionary *)param
                     channelName:(NSString *)aChannelName
                          result:(FlutterResult)result
{
    __weak typeof(self)weakSelf = self;
    BOOL requireDeliveryAck = [param[@"requireDeliveryAck"] boolValue];
    EMClient.sharedClient.options.enableDeliveryAck = requireDeliveryAck;
    [weakSelf wrapperCallBack:result
                  channelName:aChannelName
                        error:nil
                       object:nil];
}

- (void)updateSortMessageByServerTimeSetting:(NSDictionary *)param
                                 channelName:(NSString *)aChannelName
                                      result:(FlutterResult)result
{
    __weak typeof(self)weakSelf = self;
    BOOL sortMessageByServerTime = [param[@"sortMessageByServerTime"] boolValue];
    EMClient.sharedClient.options.sortMessageByServerTime = sortMessageByServerTime;
    [weakSelf wrapperCallBack:result
                  channelName:aChannelName
                        error:nil
                       object:nil];
}


@end
